### PR TITLE
feat(echarts): Implement stream graph for Echarts Timeseries

### PR DESCRIPTION
**Pull request by kgabryje**
### SUMMARY
Echarts don't support stream graph for area chart. This PR implements stream graph for Echarts Area and other Timeseries charts to achieve  feature parity with legacy NVD3 charts.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
Stream graph area chart:
image

image

Stream graph bar chart:
image

Stream graph area chart with categorical x axis:
image

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [x] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [x] Migration is atomic, supports rollback & is backwards-compatible
  - [x] Confirm DB migration upgrade and downgrade tested
  - [x] Runtime estimates and downtime expectations provided
  Current: 0.21 s
  10+: 0.12 s
  100+: 0.11 s
  1000+: 0.12 s
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by villebro** [comment]
image

File: superset-frontend/plugins/plugin-chart-echarts/src/Timeseries/transformers.ts [hunk 1]
```diff
@@ -64,11 +68,78 @@ import {
 } from '../utils/annotation';
 import { currentSeries, getChartPadding } from '../utils/series';
 import {
-  AreaChartExtraControlsValue,
   OpacityEnum,
+  StackControlsValue,
   TIMESERIES_CONSTANTS,
 } from '../constants';
 
+// based on weighted wiggle algorithm
```
Comment (villebro) [inline superset-frontend/plugins/plugin-chart-echarts/src/Timeseries/transformers.ts:76]: can we add a reference to the source?

File: superset-frontend/plugins/plugin-chart-echarts/src/Timeseries/transformers.ts [hunk 2]
```diff
@@ -64,11 +68,78 @@ import {
...
+export const getBaselineSeriesForStream = (
+  series: [string | number, number][][],
+  seriesType: EchartsTimeseriesSeriesType,
+) => {
```
Comment (villebro) [inline superset-frontend/plugins/plugin-chart-echarts/src/Timeseries/transformers.ts:80]: Could we add a simple test case for this?

File: superset-frontend/plugins/plugin-chart-echarts/src/Timeseries/Regular/Bar/controlPanel.tsx [hunk 3]
```diff
@@ -37,10 +37,13 @@ import {
...
 import {
   legendSection,
+  onlyTotalControl,
+  percentageThresholdControl,
   richTooltipSection,
   seriesOrderSection,
-  showValueSection,
+  showValueControl,
 } from '../../../controls';
+import { BarChartExtraControlsOptions } from '../../../constants';
```
Comment (villebro) [inline superset-frontend/plugins/plugin-chart-echarts/src/Timeseries/Regular/Bar/controlPanel.tsx:46]: bycatch, many '../..' in the imports

**Review by villebro — COMMENTED** [review]
First pass comments, code looks good at first glance, but would be good to add a simple test case for the wiggle algo

**Comment by michael-s-molina** [comment]
applause

File: superset/migrations/versions/2023-03-17_13-24_b5ea9d343307_bar_chart_stack_options.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+import sqlalchemy as sa
+from alembic import op
+from sqlalchemy import and_, Column, Integer, String, Text
+from sqlalchemy.ext.declarative import declarative_base
+
+from superset import db
+
+Base = declarative_base()
+
+CHART_TYPE = "echarts_timeseries_bar"
```
Comment (villebro) [inline superset/migrations/versions/2023-03-17_13-24_b5ea9d343307_bar_chart_stack_options.py:40]: shouldn't we also migrate the other ECharts "timeseries" like plugins to use the new select-based control? Here I'm using the ECharts "Generic Chart", and the stack control is still a checkbox:
image

**Review by villebro — COMMENTED** [review]
When I created a stream bar chart, then changed the chart type to line chart, the lines went completely missing:
image

I wonder if it's because the stacking control type changed from string to boolean based? Also, I think we may need to migrate all ECharts plugins that are using the boolean based stacking control to the new string based one to avoid clashes. Optionally, we may need to add one of those form data migrators that we're using when switching between some other charts to handle the control value conversions.

**Comment by kgabryje** [comment]
Thanks for comments @villebro, updated the PR. Now all timeseries echarts support stream graph :) 
image

**Comment by kgabryje** [comment]
@villebro For some reason the updated import paths (import from `@superset-ui/plugin-chart-echarts` instead of `../../../controls`) broke the unit tests. I reverted the change to unblock the CI. We can try to resolve it in a separate PR

File: superset-frontend/plugins/plugin-chart-echarts/src/Timeseries/Area/controlPanel.tsx [hunk 5]
```diff
@@ -116,6 +117,7 @@ const config: ControlPanelConfig = {
           },
         ],
         [onlyTotalControl],
+        [percentageThresholdControl],
```
Comment (villebro) [inline superset-frontend/plugins/plugin-chart-echarts/src/Timeseries/Area/controlPanel.tsx:120]: Oh nice bycatch

File: superset-frontend/plugins/plugin-chart-echarts/src/constants.ts [hunk 6]
```diff
@@ -71,18 +71,29 @@ export enum OpacityEnum {
...
   JsonValue,
   Exclude<ReactNode, null | undefined | boolean>,
 ][] = [
   [null, t('None')],
-  [AreaChartExtraControlsValue.Stack, t('Stack')],
-  [AreaChartExtraControlsValue.Expand, t('Expand')],
+  [StackControlsValue.Stack, t('Stack')],
+  [StackControlsValue.Stream, t('Stream')],
+  [StackControlsValue.Expand, t('Expand')],
 ];
```
Comment (villebro) [inline superset-frontend/plugins/plugin-chart-echarts/src/constants.ts:97]: Maybe a nit, but this had me confused: could these be called `StackControlOptions` and `AreaChartStackControlOptions`?

**Review by villebro — APPROVED** [review]
Left three major comments that turned out to be incorrect upon closer inspection 😄  But I think that one naming nit still has merit, I think (unless it also turns out to be incorrect). Tested and works great, so LGTM with an optional nit.

**Comment by villebro** [comment]
Getting ready to kill NVD3..
image

**Comment by kgabryje** [comment]
> Getting ready to kill NVD3.. image

I'll bring the garlic!

File: superset-frontend/plugins/plugin-chart-echarts/src/constants.ts (hunk 6 above)
Comment (villebro) [reply superset-frontend/plugins/plugin-chart-echarts/src/constants.ts:97]: While we're at it, could we spread the former into the latter to DRY it up?

**Comment by stephenLYZ** [comment]
OMG. It's amazing!

**Comment by rusackas** [comment]
LOVE this :) Nice work!

**Pull request merged**